src/DssProxyActionsCharter.sol

```diff
@@ -0,0 +1,696 @@
+        VatLike(vat).hope(charter);
+        uint256 wadC = _convertTo18(gemJoin, amtC);
+        // Paybacks debt to the CDP and unlocks token amount from it
+        _frob(charter, ilk, -_toInt256(wadC), -int256(art));
+        // Denies charter to access to proxy's DAI balance in the vat after execution
+        VatLike(vat).nope(charter);
+        // Exits token amount to the user's wallet as a token
+        CharterLike(charter).exit(gemJoin, msg.sender, amtC);
+    }
+}
+
+contract DssProxyActionsEndCharter is Common {
+    // Internal functions
+
+    function _free(
+        address charter,
+        address end,
+        bytes32 ilk
+    ) internal returns (uint256 ink) {
+        VatLike vat = VatLike(CharterLike(charter).vat());
+        address urp = CharterLike(charter).getOrCreateProxy(address(this));
+        uint256 art;
+        (ink, art) = vat.urns(ilk, urp);
+
+        // If CDP still has debt, it needs to be paid
+        if (art > 0) {
+            EndLike(end).skim(ilk, urp);
+            (ink,) = vat.urns(ilk, urp);
+        }
+        // Approves the charter to transfer the position to proxy's address in the vat
+        vat.hope(charter);
+        // Transfers position from CDP to the proxy address
+        CharterLike(charter).quit(ilk, address(this));
+        // Denies charter to access to proxy's DAI balance in the vat after execution
+        vat.nope(charter);
+        // Frees the position and recovers the collateral in the vat registry
+        EndLike(end).free(ilk);
+    }
+
+    // Public functions
+    function freeETH(
+        address charter,
+        address ethJoin,
+        address end
+    ) external {
+        bytes32 ilk = GemJoinLike(ethJoin).ilk();
+
+        // Frees the position through the end contract
+        uint256 wad = _free(charter, end, ilk);
+        // Fluxs to the proxy's manager proxy, so it can be pulled out with the managed gem join
+        VatLike(CharterLike(charter).vat()).flux(
+            ilk,
+            address(this),
+            CharterLike(charter).getOrCreateProxy(address(this)),
+            wad
+        );
+        // Exits WETH amount to proxy address as a token
+        CharterLike(charter).exit(ethJoin, address(this), wad);
+        // Converts WETH to ETH
+        GemJoinLike(ethJoin).gem().withdraw(wad);
+        // Sends ETH back to the user's wallet
+        msg.sender.transfer(wad);
+    }
+
+    function freeGem(
+        address charter,
+        address gemJoin,
+        address end
+    ) external {
+        bytes32 ilk = GemJoinLike(gemJoin).ilk();
+
+        // Frees the position through the end contract
+        uint256 wad = _free(charter, end, ilk);
+        // Fluxs to the proxy's manager proxy, so it can be pulled out with the managed gem join
+        VatLike(DaiJoinLike(gemJoin).vat()).flux(
+            ilk,
+            address(this),
+            CharterLike(charter).getOrCreateProxy(address(this)),
+            wad
+        );
```

Comment:
Same here, which btw gets the `vat` differently against the `freeETH` one, but should be able to use the exact same code.

Reply:
Done.